pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol

```diff
@@ -921,6 +937,117 @@ abstract contract ManagedPoolSettings is BasePool, ProtocolFeeCache, ReentrancyG
+    function getCurrentCircuitBreakerBounds(IERC20 token) public view returns (uint256, uint256) {
+        return
+            CircuitBreakerLib.getCurrentCircuitBreakerBounds(
+                _circuitBreakerState[token],
+                _getNormalizedWeight(token).complement()
+            );
+    }
+
+    /**
+     * @notice Returns flags indicating whether the lower or upper bound circuit breakers have tripped.
+     */
+    function hasCircuitBreakerTripped(IERC20 token) external view returns (bool, bool) {
+        return
+            CircuitBreakerLib.hasCircuitBreakerTripped(
+                _circuitBreakerState[token],
+                getActualSupply(),
+                _getNormalizedWeight(token),
+                _getUpscaledTokenBalance(token)
+            );
+    }
+
+    /**
+     * @notice Set a circuit breaker for one or more tokens.
+     * @dev This is a permissioned function, and disabled if the pool is paused. The lower and upper bounds
+     * are percentages, corresponding to a *relative* change in the token's spot price: e.g., a lower bound
+     * of 0.8 means the breaker should prevent trades that result in the value of the token dropping 20% or
+     * more relative to the rest of the pool.
+     */
+    function setCircuitBreakers(
+        IERC20[] memory tokens,
+        uint256[] memory lowerBoundPercentages,
+        uint256[] memory upperBoundPercentages
+    ) external authenticate whenNotPaused {
+        InputHelpers.ensureInputLengthMatch(tokens.length, lowerBoundPercentages.length, upperBoundPercentages.length);
+
+        for (uint256 i = 0; i < tokens.length; i++) {
+            _setCircuitBreaker(tokens[i], lowerBoundPercentages[i], upperBoundPercentages[i]);
+        }
+    }
+
+    // Compute the reference values, then pass them along with the bounds to the library.
+    function _setCircuitBreaker(
+        IERC20 token,
+        uint256 lowerBoundPercentage,
+        uint256 upperBoundPercentage
+    ) private {
+        // Do not allow setting a circuit breaker on the BPT itself
+        _require(address(token) != address(this), Errors.INVALID_TOKEN);
+
+        uint256 normalizedWeight = _getNormalizedWeight(token);
+
+        // Note that `getBptPrice` will revert if the token is invalid.
+        uint256 bptPrice = getBptPrice(token, normalizedWeight);
+
+        // The library will validate the lower/upper bounds
+        _circuitBreakerState[token] = CircuitBreakerLib.setCircuitBreakerFields(
+            bptPrice,
+            normalizedWeight.complement(),
+            lowerBoundPercentage,
+            upperBoundPercentage
+        );
+
+        emit CircuitBreakerSet(token, bptPrice, lowerBoundPercentage, upperBoundPercentage);
+    }
+
+    /**
+     * @notice Get the price of a single token in terms of BPT, given the weight.
+     * @dev Returns an 18-decimal floating point number.
+     */
+    function getBptPrice(IERC20 token, uint256 normalizedWeight) public view returns (uint256) {
```

Comment:
Why receive `weight`? We know that token's weight, there's not much of a point in this being an argument.